File: leetcode/JavaScript/No152.maximum-product-subarray.js
/**
 * Difficulty:
 * Medium
 *
 * Desc:
 * Find the contiguous subarray within an array (containing at least one number)
 * which has the largest product.
 * For example, given the array [2,3,-2,4],
 * the contiguous subarray [2,3] has the largest product = 6.
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
  if (!nums.length) return null
  let maxPre = nums[0]
  let minPre = nums[0]
  let max = nums[0]

  for (let i = 1; i < nums.length; i += 1) {
    const num = nums[i]
    const maxhere = Math.max(Math.max(maxPre * num, minPre * num), num)
    const minhere = Math.min(Math.min(maxPre * num, minPre * num), num)
    max = Math.max(max, maxhere)
    maxPre = maxhere
    minPre = minhere
  }
  return max
}

/**
 * @param {number[]} nums
 * @return {number}
 * https://leetcode.com/problems/maximum-product-subarray/discuss/48230/Possibly-simplest-solution-with-O(n)-time-complexity
 */
var maxProduct_2 = function(nums) {
  if (!nums.length) return null
  let maxTmp = minTmp = max = nums[0]

  for (let i = 1; i < nums.length; i += 1) {
    const num = nums[i]
    if (num < 0) {
      const tmp = maxTmp
      maxTmp = minTmp
      minTmp = tmp
    }
    maxTmp = Math.max(maxTmp * num, num)
    minTmp = Math.min(minTmp * num, num)

    max = Math.max(max, maxTmp)
  }
  return max
}

File: leetcode/C#/No152.maximum-product-subarray.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given an integer array nums, find the contiguous subarray within an array (containing at least one number) which has the largest product.
 *
 * Example 1:
 * Input: [2,3,-2,4]
 * Output: 6
 * Explanation: [2,3] has the largest product 6.
 *
 * Example 2:
 * Input: [-2,0,-1]
 * Output: 0
 * Explanation: The result cannot be 2, because [-2,-1] is not a subarray.
*/

public class Solution {
    public int MaxProduct(int[] nums) {
        if (nums.Length == 0) return 0;

        int min = nums[0];
        int max = nums[0];
        int res = nums[0];

        for (int i = 1; i < nums.Length; i += 1) {
            if (nums[i] < 0) {
                int tmp = max;
                max = min;
                min = tmp;
            }
            max = Math.Max(nums[i], max * nums[i]);
            min = Math.Min(nums[i], min * nums[i]);
            res = Math.Max(res, max);
        }
        return res;
    }
}